extensions/amp-subscriptions-google/0.1/amp-subscriptions-google.js

```diff
@@ -571,70 +617,205 @@ export class GoogleSubscriptionsPlatform {
+
+          entitlementsParams.metering = {state: meteringState};
+
+          // Remember we requested metering entitlements.
+          // This helps avoid redundant fetches for metering entitlements.
+          this.serviceAdapter_.rememberMeteringEntitlementsWereFetched();
         }
-        swgEntitlements.ack();
-        return new Entitlement({
-          source: swgEntitlement.source,
-          raw: swgEntitlements.raw,
-          service: PLATFORM_ID,
-          granted,
-          // if it's granted it must be a subscriber
-          grantReason: granted ? GrantReason.SUBSCRIBER : null,
-          dataObject: swgEntitlement.json(),
-          decryptedDocumentKey: swgEntitlements.decryptedDocumentKey,
-        });
+
+        return this.runtime_
+          .getEntitlements(entitlementsParams)
+          .then((swgEntitlements) =>
+            this.createAmpEntitlement(swgEntitlements)
+          );
       });
     });
   }
 
+  /**
+   * Returns an AMP entitlement based on SwG entitlements.
+   * @param {!Entitlements} swgEntitlements
+   * @return {Entitlement}
+   */
+  createAmpEntitlement(swgEntitlements) {
+    // Get and store the isReadyToPay signal which is independent of
+    // any entitlments existing.
+    if (swgEntitlements.isReadyToPay) {
+      this.isReadyToPay_ = true;
+    }
+
+    let swgEntitlement = swgEntitlements.getEntitlementForThis();
+    let granted = false;
+    if (swgEntitlement && swgEntitlement.source) {
+      granted = true;
+    } else if (
+      swgEntitlements.entitlements.length &&
+      swgEntitlements.entitlements[0].products.length
+    ) {
+      // We didn't find a grant so see if there is a non granting
+      // and return that. Note if we start returning multiple non
+      // granting we'll need to refactor to handle returning an
+      // array of Entitlement objects.
+      // #TODO(jpettitt) - refactor to handle multi entitlement case
+      swgEntitlement = swgEntitlements.entitlements[0];
+    } else {
+      return null;
+    }
+    swgEntitlements.ack();
+
+    // Determine grant reason.
+    let grantReason;
+    if (granted) {
+      if (swgEntitlement.source === 'google:metering') {
+        grantReason = GrantReason.METERING;
+      } else {
+        grantReason = GrantReason.SUBSCRIBER;
+      }
+    } else {
+      grantReason = null;
+    }
+
+    return new Entitlement({
+      source: swgEntitlement.source,
+      raw: swgEntitlements.raw,
+      service: PLATFORM_KEY,
+      granted,
+      // if it's granted it must be a subscriber
+      grantReason,
+      dataObject: swgEntitlement.json(),
+      decryptedDocumentKey: swgEntitlements.decryptedDocumentKey,
+    });
+  }
+
   /** @override */
   getPlatformKey() {
-    return PLATFORM_ID;
+    return PLATFORM_KEY;
   }
 
   /** @override */
-  activate(entitlement, grantEntitlement) {
+  activate(entitlement, grantEntitlement, continueAuthorizationFlow) {
     const best = grantEntitlement || entitlement;
```

Comment:
Not sure if it would be here or not but do we need to add `setShowcaseEntitlement` functionality somewhere?

Reply:
I think this would include adding `isRegistered` or similar to the `authorizationUrl` response

Reply:
Sorry, I'm not quite following, is this for analytics?

CC @jpettitt & @elijahsoria

Reply:
It's so publishers can let us know why they didn't utilize extended access. See here: https://developers.google.com/news/subscribe/extended-access/integration-steps#example-client-side-paywalls

Reply:
Thanks for explaining that!

@jpettitt , should we add this for AMP?

Reply:
Do we actually know all the values for setShowcaseEmntitlment in the AMP context? - where we do it makes sense to set it.

Reply:
My 2 cents...I think we have most of it from the pub's auth response

For example `granted && grantReason == SUBSCRIBER` should send:

```
subscriptions.setShowcaseEntitlement({
  entitlement: 'EVENT_SHOWCASE_UNLOCKED_BY_SUBSCRIPTION',
  isUserRegister: true
});
```

However, for most of the other blocks, we need to know if the user is registered which isn't obvious when `grantReason` isn't `SUBSCRIBER`. This could probably come from the auth response as well but I don't know if it would be better to update the spec to include a `isUserRegistered` boolean or to rely on an unstructured `data.isUserRegistered` response